### ISSUE TITLE
Out of memory with many gcov files

### ISSUE THREAD

**Issue by wsnyder** [RPGillespie6/fastcov#43]
First, thanks for this project.  I have a project with about 5000 coverage files to merge.  lcov is taking about 45 minutes. They all are mesauring a common ~100 source files so the final output isn't too big.

Running fastcov on about 10% of the project was almost 10x faster than lcov, so thanks!

Unfortunately I can't run it on all of the files as it runs out of memory.  
Line 113

```
        for f in intermediate_json_files:
            files.append(f) #thread safe, there might be a better way to do this though
``` 

This requires all json files to fit cumulatively in memory.  I suppose I can work on chunking my calls into fastcov, but chunking is something fastcov already seems to understand. Might you consider having processGcdas when it finishes a chunk, merging the chunk's data, clearing the intermediate json, then working on the next chunk?

Thanks

**Comment by RPGillespie6** [comment]
Wow, I figured with most computers having 4-8 GB RAM, it would be near impossible to have a project big enough to fill that up with python dicts... I guess I was wrong!

We can definitely improve this. Just out of curiosity, how much RAM does your setup have? Do you also have any idea what the value of `GCOVS_TOTAL` is at the moment of OOM?

**Comment by wsnyder** [comment]
Died at 23GB process space, with 33,000 in GCOVS_TOTAL.   If what you are looking to see is chunk size, I'd say 2000 or so would be reasonable.  I stopped it at 2000 and the main python process had 30 seconds (not including the gcov threads) and ~1.6 GB so that does seem a good chunk-o-work.

**Comment by RPGillespie6** [comment]
23GB is unacceptably huge.

To fix this, I think the easiest way will be for each worker thread to maintain a "base" fastcov json, and then for each gcov json, the thread immediately converts to fastcov json and combines with the base using the functions `distillReport` and `combineReports`. Then, we do one final combine operation after the workers are done. This will require a small amount of refactoring to get the right arguments passed in, but it will be *much* more memory efficient - there will be a max of `2*[num cores]` dictionaries in memory at any given moment vs. 33k+! This also _may_ speed up fastcov slightly, since each core is incrementally performing combine operations instead of a single core doing all combine operations later.

I can take a crack at it today or tomorrow, stay tuned (also, I noticed the fork - feel free to open PR if you are chomping at the bit)

**Comment by wsnyder** [comment]
Thanks, that would be excellent.  I have a few unrelated trivial things I'm intending to make pull requests for.

**Comment by RPGillespie6** [comment]
I have created an initial fix on the branch [bgillesp/reduce_memory_footprint](https://github.com/RPGillespie6/fastcov/tree/bgillesp/reduce_memory_footprint). This should drastically reduce memory footprint. I need to do a bit more testing before I merge to master, but branch should be fully functional.

**Comment by RPGillespie6** [comment]
@wsnyder Can you confirm the latest fastcov.py on master resolves the memory issue? If so I will tag `v1.6`. Otherwise I will re-open this issue. You should also see a ~15% speed increase.

**Comment by wsnyder** [comment]
Works perfectly - so fast thought it was broken :)
